docs/chymyst07.md

```diff
@@ -0,0 +1,731 @@
+
+```scala
+val beginCritical = b[Unit, M[Unit]]
+val access = m[Unit]
+
+site (
+  go { case beginCritical(_, reply) + access(_) =>
+    val endCritical = m[Unit] // Declare a new emitter.
+    site (
+      go { case endCritical(_) => ??? }
+    )
+    reply(endCritical) // beginCritical() returns the new emitter.
+  }
+)
+access() // Emit only one copy.
+
+```
+
+Since `endCritical` and its reaction site are defined within the local scope of the reaction that consumes `beginCritical()`, each such reaction will create a _chemically unique_ new molecule that no other reactions can emit.
+This will guarantee that reactions cannot end the critical section for other reactions.
+
+Also, since the `endCritical` emitter is created by `beginCritical()`, we cannot possibly call `endCritical()` before calling `beginCritical()`!
+So far, so good.
+
+It remains to make `endCritical()` do something useful when called.
+The obvious thing is to make it emit `access()`.
+The presence of `access()` in the soup will restore the ability of other reactions to enter the critical section.
+
+The code then looks like this:
+
+```scala
+val beginCritical = b[Unit, M[Unit]]
+val access = m[Unit]
+
+site (
+  go { case beginCritical(_, reply) + access(_) =>
+    val endCritical = m[Unit] // Declare a new emitter.
+    site (
+      go { case endCritical(_) => access() }
+    )
+    reply(endCritical) // beginCritical() returns the new emitter.
+  }
+)
+access() // Emit only one copy.
+
+// Example usage:
+val endCritical = beginCritical()
+
+???... // The code of the critical section.
+
+endCritical() // End of the critical section.
+
+```
+
+This implementation works but has a drawback: the user can call `endCritical()` multiple times.
+This will emit multiple `access()` molecules and break the logic of the critical section functionality.
+
+To fix this problem, let is think about how we could prevent `endCritical()` from starting its reaction after the first time it did so.
+The only way to prevent a reaction from starting is to omit an input molecule.
+Therefore, we need to introduce another molecule (say, `beganOnce`) as input into that reaction.
+The reaction will consume that molecule and never emit it again.
+
+```scala
+val beginCritical = b[Unit, M[Unit]]
+val access = m[Unit]
+
+site (
+  go { case beginCritical(_, reply) + access(_) =>
+    val endCritical = m[Unit] // Declare a new emitter.
+    val beganOnce = m[Unit]
+    site (
+      go { case endCritical(_) + beganOnce(_) => access() }
+    )
+    beganOnce() // Emit only one copy.
+    reply(endCritical) // beginCritical() returns the new emitter.
+  }
+)
+access() // Emit only one copy.
+
+// Example usage:
+val endCritical = beginCritical()
+
+???... // The code of the critical section.
+
+endCritical() // End of the critical section.
+
+endCritical() // This has no effect because `beganOnce()` is not available any more.
```

Comment:
No, I don't have any tests yet for this ;) You are welcome to make a PR with a new test if you feel like it. I'm not sure if you could do something with property checking here. I am not actually sure how to test the fact that critical sections work as advertised, in the complicated case displayed here.